### ISSUE TITLE
Feature request for user organization membership

### ISSUE THREAD

**Issue by pickgr** [buildkite/terraform-provider-buildkite#462]
Hello,

I see that this provider supports managing Buildkite Teams and Buildkite user memberships to those teams.  ie:

https://github.com/buildkite/terraform-provider-buildkite/blob/main/docs/resources/team_member.md

However, I don't see a resource for managing user membership to the Buildkite organization? This would be analogous to this resource in the github provider

https://registry.terraform.io/providers/integrations/github/latest/docs/resources/membership

Would this be possible? It would greatly simplify our onboarding/offboarding of users.

**Comment by james2791** [comment]
Hey @pickgr!

Hope you're well - thanks for the request too! As of now, the provider doesn't have that functionality in place as you've seen, though team and team member operations are in place for when users are added to the organization beforehand.

As you've most likely seen, the provider's interaction in place is powered by our APIs - and we do have [Update](https://buildkite.com/docs/apis/graphql/schemas/mutation/organizationmemberupdate)/[Deletion](https://buildkite.com/docs/apis/graphql/schemas/mutation/organizationmemberdelete) in place. Theres also accompanying invitation mutations, but I believe something like a `buildkite_organization_member` resource would be what you'd expect to maintain with a direct `userId` and `organizationId` to power this potential implementation?

**Comment by pickgr** [comment]
> Hey @pickgr!
> 
> Hope you're well - thanks for the request - and hope you had a great New Year to welcome in 2024!
> 
> As of now, the provider doesn't have that functionality in place as you've seen, though team and team member operations are in place for when users are added to an organization beforehand.
> 
> As you've most likely seen, the provider's interaction in place is powered by our APIs (primarily with the provider - we use the GraphQL API) - and there are [update](https://buildkite.com/docs/apis/graphql/schemas/mutation/organizationmemberupdate) and [delete](https://buildkite.com/docs/apis/graphql/schemas/mutation/organizationmemberdelete) mutations in place for organization membership _after_ users accept an [invitation](https://buildkite.com/docs/apis/graphql/cookbooks/organizations#create-a-user-add-them-to-a-team-and-set-user-permissions) to an organisation. I believe something like a `buildkite_organization_member` resource would be what you'd expect to maintain - albeit with the need to import membership after its been made (as the creation process requires that invitation/acceptance to create membership), and then let Terraform update/delete membership in state, if thats what you envision? 😃

@james2791 - thanks for the response. Yes, what you are describing sounds what I was hoping could be implemented.

**Comment by james2791** [comment]
No worries @pickgr!

We'll have a look at the potentiality of an implementation of said resource - with the known case that organization invitations do need to be accepted by users _first_ so that their membership can be statefully maintained with those (currently) two mutations.

**Comment by dattallant** [comment]
I'd like to pick this up if you'd take a PR. Proposed shape for a `buildkite_organization_member` resource, backed by the REST members endpoints (`PATCH`/`DELETE /v2/organizations/{org}/members/{uuid}`):

```hcl
resource "buildkite_organization_member" "jane" {
  email    = "jane@example.com" # or uuid
  role     = "ADMIN"            # MEMBER | ADMIN
  sso_mode = "OPTIONAL"         # REQUIRED | OPTIONAL
}
```

Create adopts an existing member (looked up by email or UUID) and errors if the user isn't a member yet, since membership itself comes from an invitation, SSO or SCIM rather than the API. Destroy only removes the member from state by default; an opt-in `remove_on_destroy = true` actually removes them from the organization. Import by UUID.

Does that match how you'd want it to behave, and would you prefer the GraphQL `organizationMemberUpdate`/`organizationMemberDelete` mutations over REST?

**Comment by scadu** [comment]
Hey @dattallant 👋 

Thanks for following up on that one and offering contribution.

We've discussed it internally and came up with the scope to add `buildkite_organization_membership` resource using REST API that:
- adopts an existing membership when found;
- optionally invites users who aren't members yet;
- tracks the invitation from pending to active membership;
- manages role and SSO mode;
- revokes pending invitations or removes active members on destroy;
- exposes user_id for use with the existing buildkite_team_member resource.

We'd leave invitation resend out of the initial version. Teams are already covered by `buildkite_team_member`. 

We plan to start working on implementation soon.

**Comment by dattallant** [comment]
that's great news @scadu , thanks - that scope covers everything we need, and exposing user_id for buildkite_team_member is a nice touch. one request: could removing an active member on destroy be opt-in (a remove_on_destroy-style flag, default off)? for orgs that adopt existing memberships, a refactor or a dropped block shouldn't be able to eject a person from the organization. happy to review or test an early version against our setup.

**Comment by scadu** [comment]
@dattallant cheers for following up! Curious to learn why you'd like to removing a membership from Terraform code be a no-op? With any resource defined in the code, removing it, would result in destroying said object. So, no-op by default would be confusing to operators. And any issues should be caught in `terraform plan`. 
References GitHub's Terraform provider follows the same logic: https://registry.terraform.io/providers/integrations/github/latest/docs/resources/membership#downgrade_on_destroy-1.

**Comment by dattallant** [comment]
fair point @scadu , and remove-on-destroy as the default is fine by me - I wasn't after a no-op, more the escape hatch the github provider has in the thing you linked (downgrade_on_destroy): for an org that adopts a few hundred existing, SSO-provisioned memberships, being able to say "on destroy, demote to member / leave alone" instead of "remove from the org" makes a refactor or a state mishap recoverable, since re-inviting a removed user doesn't bring back their teams and tokens. an opt-in flag like that would cover it; default destroy = remove.
